mmdet/datasets/pipelines/copy_paste_aug.py

```diff
@@ -0,0 +1,195 @@
+
+    @staticmethod
+    def cast(results, bbox_type, mask_type, label_type):
+        results['gt_masks'].masks = results['gt_masks'].masks.astype(mask_type)
+        results['gt_bboxes'] = results['gt_bboxes'].astype(bbox_type)
+        results['gt_labels'] = results['gt_labels'].astype(label_type)
+
+    @staticmethod
+    def extract_bboxes(results):
+        bboxes = []
+        for mask in results['gt_masks'].masks:
+            xindices = np.where(np.any(mask, axis=0))[0]
+            yindices = np.where(np.any(mask, axis=1))[0]
+            if yindices.shape[0]:
+                x1, x2 = xindices[[0, -1]]
+                y1, y2 = yindices[[0, -1]]
+            else:
+                x1, y1, x2, y2 = 0, 0, 0, 0
+            bboxes.append((x1, y1, x2, y2))
+        results['gt_bboxes'] = np.array(bboxes)
+
+    def visualize(self, results, name='', draw_ignore=False):
+        def draw_rect(img, bbox, color):
+            bbox = [int(x) for x in bbox]
+            x0, y0, x1, y1 = bbox
+            cv2.rectangle(img, (x0, y0), (x1, y1), color, 2)
+
+        img = np.copy(results['img'])
+        n = max(results['gt_bboxes'].shape[0], results['gt_masks'].masks.shape[0])
+        if draw_ignore:
+            for bbox in results['gt_bboxes_ignore']:
+                draw_rect(img, bbox, (0, 0, 255))
+        for i in range(n):
+            bbox = results['gt_bboxes'][i] if i < results['gt_bboxes'].shape[0] else None
+            mask = results['gt_masks'].masks[i] if i < results['gt_masks'].masks.shape[0] else None
+            if bbox is not None:
+                draw_rect(img, bbox, (0, 255, 0))
+            if mask is not None:
+                color = self.colors[i % len(self.colors)]
+                colored_mask = np.stack([mask, mask, mask], axis=-1) * color
+                img = cv2.addWeighted(img, 1.0, colored_mask, 0.5, 0.0)
+        cv2.imshow(name, img)
+        cv2.waitKey(0)
+
+    def rescale(self, results, target_size):
+        h, w = target_size
+        image_size = results['img'].shape[:-1]
+        if image_size != target_size:
+            if self.resize is None:
+                self.resize = Resize(img_scale=(w, h), keep_ratio=False, multiscale_mode='value')
+            else:
+                self.resize.img_scale = (w, h)
+            results['keep_ratio'] = False
+            results['scale'] = (w, h)
+            results.pop('scale_factor')
+            results = self.resize(results)
+        return results
+
+    def get_composed_mask(self, results, h, w, inds=None):
+        composed_mask = np.zeros((h, w), dtype=np.uint8)
+        for i, mask in enumerate(results['copy_paste']['gt_masks'].masks):
+            if inds is None or i in inds:
+                composed_mask = np.logical_or(composed_mask, mask)
+        return composed_mask
+
+    def __call__(self, results):
+        if not 'copy_paste' in results:
```

Comment:
`not in` is more pythonic, but this one is Ok as well.

Reply:
Fixed.